### DISCUSSION TITLE
ShadowRealm for a window context?

### DISCUSSION THREAD

**Discussion by tomvanenckevort** [sebastienros/jint#1784] [Q&A]
There seem to be a few older issues and [discussions](https://github.com/sebastienros/jint/pull/789#issuecomment-845152667) where it's mentioned that ShadowRealms might be a solution to work around the fact that v3 no longer allows you to create wrapped lexical environments. Is that still the case?

The reason I'm asking is because I'm looking into the [migration](https://github.com/AngleSharp/AngleSharp.Js/issues/84) of AngleSharp.Js to v3, which relies on a global `window` object. In the old code it would enter a new execution context with the "wrapped" lexicals, but that is no longer available.

The only hacky way I've sort of seen working so far is to wrap the script inside a `with` statement (e.g. `engine.Evaluate($"with (window) {{ {source} }}")`), since that internally still uses the [same approach](https://github.com/sebastienros/jint/blob/4fe84ea436804a9f7eeedb00cc62d706d61f22c6/Jint/Runtime/Interpreter/Statements/JintWithStatement.cs#L27-L48). But I don't think utilising that deprecated JS feature is any good.

So far I haven't been able to find any real good example uses of the ShadowRealm that show how to pass in the right context objects in the right places.

**Comment by lahma** [comment]
I actually tried to check [how AngleSharp would integrate with Jint 3.x](https://github.com/lahma/AngleSharp.Js/commit/99fa88449de3f19ef31f49eff7e955941574e8a3) among many other projects, but AngleSharp was most definitely the trickiest.  

So what I understood from the spec it would probably mean that AngleSharp would create it's own Host which there *preliminary* support for in Jint. Host would allow creating own global and setup it as wished, then engine instance would be kind of a `Window` instance. There's probably things missing and extension points not public/sealed - so this would probably need quite some investigation...
